kivy/core/image/_img_sdl2.pyx

```diff
@@ -42,53 +42,58 @@ def save(filename, w, h, fmt, pixels, flipped):
 cdef load_from_surface(SDL_Surface *image):
     cdef SDL_Surface *image2 = NULL
     cdef SDL_Surface *fimage = NULL
-    cdef SDL_PixelFormat pf
+    cdef int want_rgba = 0, target_fmt = 0, n = 0
     cdef bytes pixels
 
+    if image == NULL:
+        Logger.warn('ImageSDL2: load_from_surface() with NULL surface')
+        return None
+
+    # SDL 2.0.5 now has endian-agnostic 32-bit pixel formats like RGB24,
+    # but we can't count on that yet (RGB24 is available since 2.0)
+    if SDL_BYTEORDER == SDL_BIG_ENDIAN:
+        want_rgba = SDL_PIXELFORMAT_RGBA8888
+    else:
+        want_rgba = SDL_PIXELFORMAT_ABGR8888
+
+    # Determine if we need pixel format conversion
+    if image.format.format == want_rgba:
+        fmt = 'rgba'
+    elif image.format.format == SDL_PIXELFORMAT_RGB24:
+        fmt = 'rgb'
+    elif image.format.Amask:
+        fmt = 'rgba'
+        target_fmt = want_rgba
+    elif image.format.palette != NULL:
+        # We want RGB24 here, and rgba only if the palette has alpha
+        fmt = 'rgba'
+        target_fmt = want_rgba
+# FIXME: ... but this breaks compile, "SDL_Color has no member unused"
+#        for n in xrange(0, image.format.palette.ncolors):
+#            if image.format.palette.colors[n].a < 0xFF:
+#                fmt = 'rgba'
+#                target_fmt = want_rgba
+#                break
+    else:
+        fmt = 'rgb'
+        target_fmt = SDL_PIXELFORMAT_RGB24
+
+    # Convert if needed, and return a copy of the raw pixel data
     try:
-        if image == NULL:
-            return None
-
-        fmt = ''
-        if image.format.BytesPerPixel == 3:
-            fmt = 'rgb'
-        elif image.format.BytesPerPixel == 4:
-            fmt = 'rgba'
-
-        # FIXME the format might be 3 or 4, but it doesn't mean it's rgb/rgba.
-        # It could be argb, bgra etc. it needs to be detected correctly. I guess
-        # we could even let the original pass, bgra / argb support exists in
-        # some opengl card.
-
-        if fmt not in ('rgb', 'rgba'):
-            if fmt == 'rgb':
-                pf.format = SDL_PIXELFORMAT_BGR888
-                fmt = 'rgb'
-            else:
-                pf.format = SDL_PIXELFORMAT_ABGR8888
-                fmt = 'rgba'
-
-            image2 = SDL_ConvertSurfaceFormat(image, pf.format, 0)
+        fimage = image
+        if target_fmt != 0:
+            image2 = SDL_ConvertSurfaceFormat(image, target_fmt, 0)
             if image2 == NULL:
-                return
-
-            fimage = image2
-        else:
-            if (image.format.Rshift > image.format.Bshift):
```

Comment:
> And please readd the fixme to have a comment/fixme for hardware that support natively argb/bgra, in case of for later contribution.

I added bgr+bgra (supported by ImageData), and a comment that I think covers this. It doesn't seem like a FIXME with the format detection now (hopefully) fixed, so I went with NOTE. 

> Which is exactly the purpose of the R>B, to test the order i think (and not little/big).

Agree, seems most likely, and I think the pr is "more correct" for handling this as well as endian-ness. Quick look at the current implementation:

* All < 24bpp formats are converted to ABGR8888
  * We want RGBA8888 on big-endian here, right?
  * [See the ***32 aliases](https://github.com/spurious/SDL-mirror/blob/release-2.0.5/include/SDL_pixels.h#L265-L275) which needs a recent sdl2.

* 24/32bpp formats are returned as-is, with format 'rgb' or 'rgba', **unless** Rshift>Bshift is true.
  * Example 1: ARGB2101010 is incorrectly returned as 'rgba' (unless the Rshift>Bshift test is true, in which case it is converted to ABGR8888 - probably, see below)
  * Example 2: RGB888 (which is **32-bit** SDL_PACKEDORDER_XRGB) will be returned as 'rgba'. Not 100% sure how that works out byte-wise, but seems broken

So about the Rshift>Bshift test:

* When true the pixel format is converted via custom SDL_PixelFormat. As I pointed out above, I am confused about this approach.
  * What are the consequences of pf.[RGBA]shift = 0? Of BytesPerPixel = 0? Is it luck or planning that the called function gives the desired result in this case??
  * Seems like it will convert to ABGR8888 in practice (unless the 
custom pixel format serves a purpose I don't understand)
  
The update should do:

* 24-bit RGB and BGR used as-is (RGB24, BGR24)
* 32-bit RGBA and BGRA used as-is (want_rgba, want_bgra)
* Everything else is converted to want_rgba (if alpha) or RGB24 (no alpha)